module/validation/seal_validator.go

```diff
@@ -0,0 +1,261 @@
+			// ExecutionResult. We are traversing the fork backwards, so we can
+			// overwrite any previously recorded result.
+
+			// ATTENTION:
+			// Here, IncorporatedBlockID (the first argument) should be set
+			// to ancestorID, because that is the block that contains the
+			// ExecutionResult. However, in phase 2 of the sealing roadmap,
+			// we are still using a temporary sealing logic where the
+			// IncorporatedBlockID is expected to be the result's block ID.
+			incorporatedResults[resultID] = flow.NewIncorporatedResult(
+				receipt.ExecutionResult.BlockID,
+				&receipt.ExecutionResult,
+			)
+
+		}
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// Loop forward across the fork and try to create a valid chain of seals.
+	// blockIDs, as populated by the previous loop, is in reverse order.
+	for i := len(blockIDs) - 1; i >= 0; i-- {
+		// if there are no more seals left, we can exit earlier
+		if len(byBlock) == 0 {
+			return last, nil
+		}
+
+		// return an error if there are still seals to consider, but they break
+		// the chain
+		blockID := blockIDs[i]
+		seal, found := byBlock[blockID]
+		if !found {
+			return nil, engine.NewInvalidInputErrorf("chain of seals broken (missing: %x)", blockID)
+		}
+
+		delete(byBlock, blockID)
+
+		// check if we have an incorporatedResult for this seal
+		incorporatedResult, ok := incorporatedResults[seal.ResultID]
+		if !ok {
+			return nil, engine.NewInvalidInputErrorf("seal %x does not correspond to a result on this fork", seal.ID())
+		}
+
+		// check the integrity of the seal
+		err := s.validateSeal(seal, incorporatedResult.Result)
+		if err != nil {
+			if engine.IsInvalidInputError(err) {
+				return nil, fmt.Errorf("payload includes invalid seal (%x), %w", seal.ID(), err)
+			}
+
+			return nil, fmt.Errorf("unexpected seal validation error %w", err)
+		}
+
+		last = seal
+	}
+
+	// at this point no seals should be left
+	if len(byBlock) > 0 {
+		return nil, engine.NewInvalidInputErrorf("not all seals connected to state (left: %d)", len(byBlock))
+	}
+
+	return last, nil
+}
+
+// validateSeal performs integrity checks of single seal. To be valid, we
+// require that seal:
+// 1) Contains correct number of approval signatures, one aggregated sig for each chunk.
+// 2) Every aggregated signature contains valid signer ids. module.ChunkAssigner is used to perform this check.
+// 3) Every aggregated signature contains valid signatures.
+// Returns:
+// * nil - in case of success
+// * engine.InvalidInputError - in case of malformed seal
+// * exception - in case of unexpected error
+func (s *sealValidator) validateSeal(seal *flow.Seal, executionResult *flow.ExecutionResult) error {
+	if len(seal.AggregatedApprovalSigs) != executionResult.Chunks.Len() {
+		return engine.NewInvalidInputErrorf("mismatching signatures, expected: %d, got: %d",
+			executionResult.Chunks.Len(),
+			len(seal.AggregatedApprovalSigs))
+	}
+
+	assignments, err := s.assigner.Assign(executionResult, seal.BlockID)
```

Comment:
I know about this, but somewhere in code I have found that this is intended behavior for now, and it will be changed later.

> // ATTENTION: For sealing phase 2, the value for IncorporatedBlockID
> // is the block the result pertains to (here parentBlock).
> // In later development phases, we will change the logic such that
> // IncorporatedBlockID references the
> // block which actually incorporates the result:

Reply:
I am not sure if we need to add it now or somewhere later

Reply:
but that has probably nothing to do with it, I've just messed up with it.